src/qt/transactiongraphwidget.cpp

```diff
@@ -0,0 +1,376 @@
+        AddToArithmeticMean(fInstantaneousTpsAverage_Sampled, vInstantaneousSamples.count(), nInstantaneousTxPerSec);
+
+    // NOTE: Instantaneous mean for displayed samples is handled in updateTransactionsPerSecondLabelValues()
+
+    // Purge instantaneous sample(s) that have moved beyond the sample set size limit
+    while (vInstantaneousSamples.size() > MAXIMUM_SAMPLES_TO_KEEP)
+    {
+        float fRemoving = vInstantaneousSamples.last();
+        // Adjust instantaneous mean for total sample set to exclude the sample being purged
+        fInstantaneousTpsAverage_Sampled =
+            SubtractFromArithmeticMean(fInstantaneousTpsAverage_Sampled, vInstantaneousSamples.size(), fRemoving);
+
+        vInstantaneousSamples.pop_back();
+    }
+
+
+    // Add new smoothed sample
+    vSmoothedSamples.push_front((float)nTxPerSec);
+
+    // Update smoothed peak for the total runtime
+    // NOTE: This peak may no longer be in the sample set if it was purged for being too old
+    if (nTxPerSec > fSmoothedTpsPeak_Runtime)
+        fSmoothedTpsPeak_Runtime = nTxPerSec;
+
+    // Adjust smoothed mean for total runtime to include new sample
+    fSmoothedTpsAverage_Runtime = AddToArithmeticMean(fSmoothedTpsAverage_Runtime, nTotalSamplesRuntime, nTxPerSec);
+
+    // Adjust smoothed mean for total sample set to include new sample
+    fSmoothedTpsAverage_Sampled = AddToArithmeticMean(fSmoothedTpsAverage_Sampled, vSmoothedSamples.count(), nTxPerSec);
+
+    // NOTE: Smoothed mean for displayed samples is handled in updateTransactionsPerSecondLabelValues()
+
+    // Purge smoothed sample(s) that have moved beyond the sample set size limit
+    while (vSmoothedSamples.size() > MAXIMUM_SAMPLES_TO_KEEP)
+    {
+        float fRemoving = vInstantaneousSamples.last();
+        // Adjust smoothed mean for total sample set to exclude the sample being purged
+        fSmoothedTpsAverage_Sampled =
+            SubtractFromArithmeticMean(fSmoothedTpsAverage_Sampled, vSmoothedSamples.size(), fRemoving);
+
+        vSmoothedSamples.pop_back();
+    }
+
+    // Update the TPS values matching the current display window
+    updateTransactionsPerSecondLabelValues();
+
+    // Limit redraw requests
+    static int64_t nLastRedrawTime = GetTimeMillis();
+    int64_t nCurrentTime = GetTimeMillis();
+    if (nCurrentTime >= nLastRedrawTime + nRedrawRateMillis)
+    {
+        update();
+        nLastRedrawTime = nCurrentTime;
+    }
+}
+
+void TransactionGraphWidget::updateTransactionsPerSecondLabelValues()
+{
+    int numSamples = getSamplesInDisplayWindow();
+
+    // Recompute the instantaneous peak for the current sample set and the average for the display window
+    // Runtime peak is computed in the add sample method because that tracks the peak even beyond the maintained sample
+    // set
+    float tSumAll = 0.0f, tSumDisplay = 0.0f;
+    int countAll = vSmoothedSamples.size();
+    int countDisplay = countAll < numSamples ? countAll : numSamples;
```

Comment:
make sure this is at least 1 to avoid div by 0 in line 316

Reply:
#1960 
Corrected by explicitly checking for > 0 in those two locations before attempting division.